cognite/client/credentials.py

```diff
@@ -59,7 +62,176 @@ def authorization_header(self) -> Tuple[str, str]:
         return "Authorization", f"Bearer {self.__token_factory()}"
 
 
-class OAuthClientCredentials(CredentialProvider):
+class _OAuthCredentialProviderWithTokenRefresh(CredentialProvider):
+    # This ensures we don't return a token which expires immediately, but within minimum 3 seconds.
+    __TOKEN_EXPIRY_LEEWAY_SECONDS = 3
```

Comment:
why 3 secs btw?

Reply:
Chosen kind of arbitrarily. Just needs to be long enough so that the token is still valid when the auth api verifies the signature.